### PR TITLE
Goal: Make it possible to move PeerPool into an isolated plugin

### PR DESCRIPTION
**Pull request by cburgdorf**
### What was wrong?

I think we are more or less aligned in our vision that we need to push forward the multiprocessing architecture that we've started. Similar to how modern browsers work (or many other applications), Trinity needs to run many processes in parallel to be able to handle the workload it is forced to deal with. This is especially true in a language such as Python that does not handle threading efficiently.

Today, we already have the [JSON-RPC Server](https://github.com/ethereum/trinity/blob/fa370299106b6391e203e8945f44f09f2c9d8703/trinity/plugins/builtin/json_rpc/plugin.py#L43), the [Peer Discovery](https://github.com/ethereum/trinity/blob/fa370299106b6391e203e8945f44f09f2c9d8703/trinity/plugins/builtin/peer_discovery/plugin.py#L144) and the [ETH Stats Reporting](https://github.com/ethereum/trinity/blob/fa370299106b6391e203e8945f44f09f2c9d8703/trinity/plugins/builtin/ethstats/plugin.py#L31) running as isolated processes (by deriving from `BaseIsolatedPlugin` which handles the heavy lifting).

Often these processes need to talk to other parts of Trinity. E.g. the `PeerPool` needs to talk to the discovery, the ETH Stats Reporting needs to ask the `PeerPool` for the number of connected nodes etc.

We leverage a central `EventBus` to handle such inter-process communication.

However, most of the networking is still running rather tightly coupled in one hardcoded (read: not managed through a plugin) process known as the `networking` process.

To be able to move core stuff such as *syncing*, *tx pool* etc into its own isolated processes, we need to get rid of the fact that most interactions with the `PeerPool` still rely on the code being run in the same process as the `PeerPool`.

In other words, what I believe, the best way forward would look like, is to have the `PeerPool` itself run as an isolated process and have all interactions be made through the `EventBus` (related: [event-bus-all-the-things](https://github.com/ethereum/trinity/issues/79), [write-an-eventbus-friendly-peerpool](https://github.com/ethereum/trinity/issues/89))

This PR is an attempt to experiment with that idea.

### How was it fixed?

Not yet. Just getting started.

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

**Comment by cburgdorf** [comment]
@jannikluhn This is basically starting what @pipermerriam pointed out with his recent [comment on your PR](https://github.com/ethereum/trinity/pull/163/files#r247567007)

So, this will most likely be related to your current work with the beacon node. I wanted to put this PR out very early on for transparency. I also tried to give some background on the direction in the description.

If you feel like doing your own experimentation on the same direction (motivated by getting the beacon node syncing), please do not feel obligated to wait for this PR. In fact, I'd be happy if someone would explore similar territory to hedge against my own failure haha.

File: p2p/events.py [hunk 1]
```diff
@@ -56,3 +57,17 @@ class ConnectToNodeCommand(BaseEvent):
 
     def __init__(self, node: str) -> None:
         self.node = node
+
+
+class ConnectedPeersResponse(BaseEvent):
+
+    # FIXME: it's going to become really hard to get rid of `Any` here :-/
```
Comment (pipermerriam) [inline p2p/events.py]: Can you expand on why?  Intuition says this would be done with `typing.TypeVar('TPeer', bound=p2p.peer.BasePeer)` but I assume there's a reason why that doesn't work.

File: trinity/protocol/bcc/peer.py [hunk 2]
```diff
@@ -41,6 +43,12 @@
 )
 
 
+class BCCPeerDTO(NamedTuple):
+    head_slot: SlotNumber
+    network_id: int
+    uri: str
```
Comment (pipermerriam) [inline trinity/protocol/bcc/peer.py]: maybe including a referencable unique identifier like the enode public key or the full enode address itself would make sense here?

File: p2p/events.py (hunk 1 above)
Comment (cburgdorf) [reply p2p/events.py]: Yes, but only if we make the `BasePeerPool` generic over `TPeer` (`BasePeerPool[TPeer]`). Actually, it's probably not going to be **that** hard as the comment suggest. I'm being overly dramatic here haha.

File: trinity/protocol/bcc/peer.py (hunk 2 above)
Comment (cburgdorf) [reply trinity/protocol/bcc/peer.py]: Isn't `uri` (coming from `Node.uri`) that unique identifier already?

Comment (pipermerriam) [reply trinity/protocol/bcc/peer.py]: quadruple-facepalm

**Comment by cburgdorf** [comment]
I created a basic benchmark to see how communication via the EventBus compares to using native Python proxies: https://github.com/ethereum/lahja/pull/29

It looks like are currently looking at a 15x poorer performance compared to native Python proxies which makes me believe that we should do the following:

1. Use proxy based access for things that are frequently needed (e.g. performing actions on peers)
2. Only expose not-too-frequent events on the EventBus so far (e.g. maybe sth like `NewChainTip`)
3. Look into ways to make the EventBus more efficient (e.g. get rid of extra process hop, use something like gRPC underneath etc)

**Comment by cburgdorf** [comment]
@pipermerriam The code is still ugly but this PR now contains a commit that moves the `ETHRequestServer` into its own isolated plugin. This already lays the ground work to do the same with the sync. I'm going to continue with that now and clean it all up once I have more confidence things will work out fine.

File: p2p/events.py [hunk 3]
```diff
@@ -56,3 +63,24 @@ class ConnectToNodeCommand(BaseEvent):
...
+class ConnectedPeersResponse(BaseEvent):
+
+    def __init__(self, peers: Tuple[IdentifiablePeer, ...]) -> None:
+        self.peers = peers
+
+
+class ConnectedPeersRequest(BaseRequestResponseEvent[ConnectedPeersResponse]):
+
+    @staticmethod
+    def expected_response_type() -> Type[ConnectedPeersResponse]:
```
Comment (pipermerriam) [inline p2p/events.py]: not for this PR but updating this method to be `get_expected_response_type` might be more inline with our other naming conventions.

**Comment by cburgdorf** [comment]
Replaced by #721
